### PR TITLE
Clues - Souvenirs - Random Temp powers in Lua

### PR DESCRIPTION
**Pull request by ZopharShinta**
Added Clues and Souvenirs to Lua.
Added Random temp power option in Lua.
Updated LuaBot to test new features.

### CODE REVIEW

File: Projects/CoX/Common/GameData/Clue.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+using vClueList = std::vector<Clue>;
+
+
+class Souvenir
+{
+public:
+     enum : uint32_t {class_version       = 1};
+
+    uint32_t m_idx;
+    void *field4;
```
Comment (broxen) [inline Projects/CoX/Common/GameData/Clue.h]: You can get rid of this. We aren't using it anywhere

File: Projects/CoX/Common/Messages/Map/ClueList.h [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    {
+        m_clue_list = clue_list;
+    }
+
+    void serializeto(BitStream &bs) const override
+    {
+        bs.StorePackedBits(1, type()-evFirstServerToClient); // packet 70
+
+        bs.StorePackedBits(1, m_clue_list.size());
+        int count = 0;
```
Comment (broxen) [inline Projects/CoX/Common/Messages/Map/ClueList.h]: You don't use `count` anywhere here, so I would remove it.

File: Projects/CoX/Common/Messages/Map/ClueList.h [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    {
+
+    public:
+        // [[ev_def:field]]
+        int32_t m_souvenir_idx = 0;
+        explicit SouvenirDetailRequest() : MapLinkEvent(MapEventTypes::evSouvenirDetailRequest){}
+
+        void serializeto(BitStream &/*bs*/) const override
+        {
+            assert(!"SouvenirDetailRequest serializeto");
```
Comment (broxen) [inline Projects/CoX/Common/Messages/Map/ClueList.h]: I would add a comment here about what packet this is (`68`). Just to clarify for future programmers reading this section of code.

File: Projects/CoX/Servers/MapServer/DataHelpers.cpp [hunk 4]
```diff
@@ -526,6 +527,18 @@ void sendStance(MapClientSession &sess, PowerStance &stance)
     sess.addCommand<SendStance>(stance);
 }
 
+void sendClueList(MapClientSession &sess)
+{
+    vClueList clue_list = sess.m_ent->m_char->m_char_data.m_clue_souvenir_list.m_clue_list;
+    sess.addCommandToSendNextUpdate(std::unique_ptr<ClueList>(new ClueList(clue_list)));
```
Comment (broxen) [inline Projects/CoX/Servers/MapServer/DataHelpers.cpp]: These you can change to the new format: `sess.addCommand<ClueList>(clue_list);`

File: Projects/CoX/Servers/MapServer/DataHelpers.cpp [hunk 5]
```diff
@@ -526,6 +527,18 @@ void sendStance(MapClientSession &sess, PowerStance &stance)
...
+}
+
+void sendSouvenirList(MapClientSession &sess)
+{
+    vSouvenirList souvenir_list = sess.m_ent->m_char->m_char_data.m_clue_souvenir_list.m_souvenir_list;
+    sess.addCommandToSendNextUpdate(std::unique_ptr<SouvenirListHeaders>(new SouvenirListHeaders(souvenir_list)));
```
Comment (broxen) [inline Projects/CoX/Servers/MapServer/DataHelpers.cpp]: Same here. Use the new syntactic sugar `addCommand<T>()`

File: Projects/CoX/Servers/MapServer/DataHelpers.cpp [hunk 6]
```diff
@@ -1222,4 +1235,26 @@ void showMapMenu(MapClientSession &sess)
     QString msg_body = createMapMenu();
     showMapXferList(sess, has_location, location, msg_body);
 }
+
+void addClue(MapClientSession &cl, Clue clue)
+{
+    vClueList clue_list = cl.m_ent->m_char->m_char_data.m_clue_souvenir_list.m_clue_list;
+    clue_list.push_back(clue);
+    cl.m_ent->m_char->m_char_data.m_clue_souvenir_list.m_clue_list = clue_list;
+    cl.addCommandToSendNextUpdate(std::unique_ptr<ClueList>(new ClueList(clue_list)));
```
Comment (broxen) [inline Projects/CoX/Servers/MapServer/DataHelpers.cpp]: Same here. Use the new syntactic sugar `addCommand<T>()`

File: Projects/CoX/Servers/MapServer/DataHelpers.cpp [hunk 7]
```diff
@@ -1222,4 +1235,26 @@ void showMapMenu(MapClientSession &sess)
...
+{
+    vSouvenirList souvenir_list = cl.m_ent->m_char->m_char_data.m_clue_souvenir_list.m_souvenir_list;
+    if (souvenir_list.size() > 0)
+        souvenir.m_idx = souvenir_list.size() + 1;
+    else
+        souvenir.m_idx = 0;
+
+    souvenir_list.push_back(souvenir);
+    cl.m_ent->m_char->m_char_data.m_clue_souvenir_list.m_souvenir_list = souvenir_list;
+    cl.addCommandToSendNextUpdate(std::unique_ptr<SouvenirListHeaders>(new SouvenirListHeaders(souvenir_list)));
```
Comment (broxen) [inline Projects/CoX/Servers/MapServer/DataHelpers.cpp]: Same here. Use the new syntactic sugar `addCommand<T>()`

File: Projects/CoX/Servers/MapServer/DataHelpers.cpp [hunk 8]
```diff
@@ -1222,4 +1235,26 @@ void showMapMenu(MapClientSession &sess)
...
+    cl.m_ent->m_char->m_char_data.m_clue_souvenir_list.m_clue_list = clue_list;
+    cl.addCommandToSendNextUpdate(std::unique_ptr<ClueList>(new ClueList(clue_list)));
+}
+
+void addSouvenir(MapClientSession &cl, Souvenir souvenir)
+{
+    vSouvenirList souvenir_list = cl.m_ent->m_char->m_char_data.m_clue_souvenir_list.m_souvenir_list;
+    if (souvenir_list.size() > 0)
+        souvenir.m_idx = souvenir_list.size() + 1;
+    else
```
Comment (broxen) [inline Projects/CoX/Servers/MapServer/DataHelpers.cpp]: if `souvenir.m_idx` is declared with an initial value of 0, you don't need this else statement

File: Projects/CoX/Servers/MapServer/MapInstance.cpp [hunk 9]
```diff
@@ -2884,4 +2892,33 @@ void MapInstance::on_trade_updated(TradeWasUpdatedMessage* ev)
...
+    }
+
+    if(!found)
+    {
+        qCDebug(logScripts) << "SouvenirDetail Souvenir " << ev->m_souvenir_idx << " not found";
+        souvenir_detail.m_idx = 0; // Should always be found?
+        souvenir_detail.m_description = "Data not found";
+
+    }
+    session.addCommandToSendNextUpdate(std::unique_ptr<SouvenirDetail>(new SouvenirDetail(souvenir_detail)));
```
Comment (broxen) [inline Projects/CoX/Servers/MapServer/MapInstance.cpp]: Same here. Use the new syntactic sugar `addCommand<T>()`

File: Projects/CoX/Common/GameData/Clue.h [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+     enum : uint32_t {class_version       = 1};
+
+    QString m_name;
+    QString m_display_name;
+    QString m_detail_text;
+    QString m_icon_file;
+
+    // for scripting language access.
+    std::string getName() const { return m_name.toStdString();}
+    void setName(const char *n) { m_name = n; }
```
Comment (nemerle) [inline Projects/CoX/Common/GameData/Clue.h]: Hmm, is the reason for using std::string here, the sol2/lua interop ?

File: Projects/CoX/Common/GameData/Clue.h [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    void setDetailText(const char *n) { m_detail_text = n; }
+
+    std::string getIconFile() const { return m_icon_file.toStdString();}
+    void setIconFile(const char *n) { m_icon_file = n; }
+
+
+    template<class Archive>
+    void serialize(Archive &archive, uint32_t const version);
+};
+using vClueList = std::vector<Clue>;
```
Comment (nemerle) [inline Projects/CoX/Common/GameData/Clue.h]: hmm, 'v' for 'vector of' and ClueList for list of Clue ? :)
Maybe we should rename this to `ClueStorage/vClues/ClueList`

Comment (dracc) [reply Projects/CoX/Common/GameData/Clue.h]: I'd prefer `ClueContainer` even. Or is the type of container used vital for this specific variable? :)

File: Projects/CoX/Common/GameData/Clue.h (hunk 10 above)
Comment (ZopharShinta) [reply Projects/CoX/Common/GameData/Clue.h]: The original example that was in place used it. So I kept it going. I know Lua doesn't know what a QString is. We can add it as a user type but we'll need to convert Lua's string to a QString before we send it.

File: Projects/CoX/Common/GameData/Clue.h (hunk 11 above)
Comment (ZopharShinta) [reply Projects/CoX/Common/GameData/Clue.h]: I could even split the clues/souvenirs out into their own objects. Doesn't matter to me. Whatever you guys feel is best.

Comment (broxen) [reply Projects/CoX/Common/GameData/Clue.h]: Seems like Clues/Souvenirs are always sent to the client together in the same array? Might not be realistic to separate into different containers.

**Pull request merged**